Implements Site eligibility check for Blaze in Publised Posts

AjeshRPai opened the pull request:
Part of #17911

## What? 
This PR adds the implementation to check whether a site is eligible of blaze in the flow Posts → Published Posts . 

In this PR I have also added  a few checks before we make an API call to fetch the Blaze status 
- Check if the User is eligible with `siteModel.isAdmin` check
- Check if the App is **Jetpack** pack and Flag is enabled

 ## How? 
- [Adds: check to see if Blaze status needs to be fetched](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/1077b7e02c8926879848b061d07cf1337face377)
- [Updates: Logic of whether the post is eligible for blaze feature](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/bee855416dd87d72b342ed74267968d5ce98344b)
- [Adds: check to see if Blaze status needs to be fetched](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/e14e5c5102b95ac7c7ea83e4fde9d977bbc6dfdd)
- [Adds: the check for blaze eligibility in post list viewmodel](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/0772cf3137a2751c3b4419e1498c25e87db938d2)


## Testing instructions

### Test - Blaze is not available

1. Go to the **Jetpack** app
2. 🚪 Login with a wp account without Administrator privileges or Use a private site 
3. ⛳ Go to **app settings** → **Debug settings →** enable feature flag → `blaze`
4. 👉🏻Go to Posts → Published Posts 
5. 👆🏻Click on the more icon ⁞ → 
6. ✅ Verify that the **🔥 Promote with Blaze** is not shown 

### Test - Blaze is available

1. Go to the **Jetpack** app
2. 🚪 Login with a wp account with Administrator privileges 
3. ⛳ Go to **app settings** → **Debug settings →** enable feature flag → `blaze`
4. 👉🏻Go to Posts → Published Posts 
5. 👆🏻Click on the more icon ⁞ → 
6. ✅ Verify that the **🔥 Promote with Blaze** is shown 

### Test - Blaze status is not fetched - Monitor on Logcat

1. Go to the **WordPress** app
2. 🚪 Login with a wp account with Administrator privileges 
3. ✅ Verify that the API is not called 
    - Open the logcat
    - Make sure that there is no occurrence of the event **`BlazeStore: fetch blaze status`**
4. ⛳ Go to **app settings** → **Debug settings →** enable feature flag → `blaze`
5. Go to the Home page 
6. ✅ Verify Step 3 again

## Regression Notes
1. Potential unintended areas of impact
- Blaze eligibility for a site is not fetched on Jetpack app and flag enabled 
- Blaze eligibility for a site is not checked on the Posts → Published Posts flow


2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests and relied on the existing Unit tests

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Adds: the check for blaze eligibility in post list viewmodel

AjeshRPai committed:
Updates: Logic of whether the post is eligible for blaze feature

AjeshRPai committed:
Adds: the logic to check whether blaze is eligible for a site

AjeshRPai committed:
Adds: check to see if Blaze status needs to be fetched

- Adds: the check to see if the app is wordpress or the
flag is enabled
- Adds: the check to see if the user has the admin privileges

AjeshRPai committed:
Adds: check to see if Blaze status needs to be fetched

AjeshRPai committed:
Updates: Logic of whether the post is eligible for blaze feature

AjeshRPai committed:
Updates: logic to use existing function

AjeshRPai committed:
Fixes: test failures and lint error

AjeshRPai committed:
Removes: Blaze eligibility check from PostListViewModel and test class

AjeshRPai committed:
Adds: the logic of checking the site eligibility to the viewmodel

Adds: the dependencies to the mock classes

AjeshRPai commented:
Hey @zwarm 👋🏼 

After testing thoroughly, I found out that the logic for checking whether a site has Blaze eligibility is not triggered before listing the Published posts. Hence no posts had the **Promote with Blaze** option in more menu. On Further investigation, I found out that there is two view models for PostListing. 😅 . So to clarify what I have done here and for my reference, I am explaining it below. Feel free to ignore if you understood with the commits or file changes. 

The current view models and the existing logic(abstracted for the context of this PR)
1. `PostListMainViewModel` - Gets initiated first, an object of `PostListViewModelConnector` is created in it 
2. `PostListViewModel` - `PostListFragment` fetches the connector from the `PostListMainViewModel` and passes to the `PostListViewModel` start method 

At first the logic for checking blaze status was in `PostListViewModel`, which I removed with the commit [be78c43](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/be78c438da60210036af7020b13e622833667786) and moved the logic to `PostListMainViewModel` with the [2cd5b96](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/2cd5b9608ef64b968c88e1d38c9874210f131693). In the same commit, I added a property in the connector for Site Blaze eligibility. [Line](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/2cd5b9608ef64b968c88e1d38c9874210f131693#diff-c2bcdfdcc1950d5c14144fcd2e163875b7759d7513d4755ac46eb8c797cf90c8R363).So `PostListMainViewModel` fetches the site eligibility and instantiates the `PostListViewModelConnector` variable with it and then `PostListFragment` fetches the connector from the `PostListMainViewModel` and passes to the `PostListViewModel` start method. 

I hope this helps clarify things.

AjeshRPai committed:
Suppress: Lint error

zwarm commented:
@AjeshRPai 
>  On Further investigation, I found out that there is two view models for PostListing. 😅 . So to clarify what I have done here and for my reference, I am explaining it below. 

I really appreciated this explanation; as I haven't delved deep into the posts list logic. 🙇

zwarm reviewed (changes requested):
@AjeshRPai - The "if" wrap in the `PromoteWithBlazeCardSource.build` method is stopping the refresh indicator from disappearing on PTR when blaze is not enabled. I believe the refresh indicator is being checked for each source, but this source never sets/resets it so it gets stuck. Can you take a look? Thanks.

AjeshRPai committed:
Fixes: infinite loading on refresh

AjeshRPai committed:
Fixes: the promote blaze card not being shown in refresh

Fixes: the promote blaze card not being shown after refresh even if
 there is eligibility

AjeshRPai commented:
Hey @zwarm 👋🏼 

Thanks for the review, I have fixed the infinite loading of the indicator with the commit [ed4ced4](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/ed4ced4eb28f7a38756eb1827bd27b013eb5616d). 

Also, while testing the infinite loading, I found a small issue with the current implementation of the Promote blaze card.

**Issue** - Once we post the `BlazeStatusModel` to the live data on the `fun build`, we are not updating it again on `refresh` → `fetchBlazeStatusAndPostErrorIfAvailable` → `model != null`. We post the same value even if we get a successful response(with or without a blaze status change) from the backend. 

To reproduce this issue from UI, you can 
1. Login with a Blaze-eligible site
2. Disable the feature flag 
3. Go to Dashboard → Verify that the Promote card is not shown 
4. Enable the feature flag 
5. Notice that the Promote card is not shown 

I have fixed this issue with the commit - [e64af0a](https://github.com/wordpress-mobile/WordPress-Android/pull/18026/commits/e64af0ab4419da5dfd714d5a6445efadaca75baa). Let me know if it makes sense.

AjeshRPai committed:
Updates: logic to use existing function

AjeshRPai committed:
Fixes: test failures

zwarm commented:
>  I have fixed the infinite loading of the indicator  

👍 

> Also, while testing the infinite loading, I found a small issue with the current implementation of the Promote blaze card.
> **Issue** - Once we post the BlazeStatusModel to the live data on the fun build, we are not updating it again on refresh → fetchBlazeStatusAndPostErrorIfAvailable → model != null. We post the same value even if we get a successful response(with or without a blaze status change) from the backend.

> I have fixed this issue 

👍 

I think this is a larger issue with ALL the "sources" and PTR. If you take a look at all the other sources they work exactly the same; no results are returned upon successful fetch. I believe the intent is that a "get" always follows the "fetch", which will then return the result as taken from the db. I've been having doubts about this logic since late last week. @thomashorta just changed the logic in this source and the blogging source to post the last state. I strongly believe that the time has come to review how this works across the board.  

Thanks.

zwarm reviewed (approved):
@AjeshRPai - Thanks for the changes. The indicator is working as expected again. 👏
